cli: added support to sparse images flashing

gfabiano opened the pull request:
A first implementation to sparse images flashing. I tested it with some images and it works correctly on my qcom bench.

gfabiano committed:
cli: added support to sparse images flashing

Signed-off-by: Giuseppe Fabiano <gfabiano40@gmail.com>

gfabiano commented:
I think that I must improve the implementation because typically raw chunks are contiguos so calling for every chunk `firehose_program_storage` can hog the speed. @quic-kdybcio what you propose, maybe moving the sparse handling inside the `firehose_program_storage` can be  better?

quic-kdybcio commented:
> I think that I must improve the implementation because typically raw chunks are contiguos so calling for every chunk `firehose_program_storage` can hog the speed. @quic-kdybcio what you propose, maybe moving the sparse handling inside the `firehose_program_storage` can be better?

I think the current approach is ok - IIUC sparse chunks can be as big as you want (which `program_storage` handles gracefully) so long as they're 4K-aligned in size.

gfabiano committed:
cli: changed the logic to speed up transfer and not hog speed

The previous logics tried to flash each useful chunk one by one
in 4KB chunks but this hogged the speed due to firehose transfers
that took ~20-50ms for each XML command ACK cycle.

Inspired by official qcom tool I merged different type of sparsed
image chunks filling also the gaps of don't care with zeroes and
try to keep large contiguous transfers. This allowed the flash to
speedup from minutes to some seconds on sparse images like persist.

Signed-off-by: Giuseppe Fabiano <gfabiano40@gmail.com>
